kernel/arch/riscv/early_fdt.c

```diff
@@ -0,0 +1,395 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+#include <tilck/common/basic_defs.h>
+#include <tilck/common/boot.h>
+#include <tilck/common/string_util.h>
+#include <tilck/common/printk.h>
+#include <tilck/common/utils.h>
+#include <tilck/kernel/errno.h>
+#include <tilck/kernel/hal.h>
+#include <3rd_party/fdt_helper.h>
+#include <libfdt.h>
+#include <multiboot.h>
+#include "paging_int.h"
+
+#define CMDLINE_BUF_SZ 1024
+
+struct simplefb_bitfield {
+   u8 offset;   /* beginning of bitfield */
+   u8 size;       /* length of bitfield */
+};
+
+struct simplefb_format {
+   const char *name;
+   u8 bpp;  /* bits per pixel */
+   struct simplefb_bitfield red;
+   struct simplefb_bitfield green;
+   struct simplefb_bitfield blue;
+   struct simplefb_bitfield alpha;
+};
+
+void *fdt_blob;   //save virtual address of fdt
+
+static ulong initrd_paddr;
+static ulong initrd_size;
+
+static multiboot_info_t *mbi;
+static multiboot_module_t *mod;
+static multiboot_memory_map_t *mmmap;
+
+static int mmmap_count = 0;
+static char *cmdline_buf;
+
+static struct simplefb_format simplefb_formats[] = {
+   {"r5g6b5", 16, {11, 5}, {5, 6}, {0, 5}, {0, 0}},
+   {"r5g5b5a1", 16, {11, 5}, {6, 5}, {1, 5}, {0, 1}},
+   {"x1r5g5b5", 16, {10, 5}, {5, 5}, {0, 5}, {0, 0}},
+   {"a1r5g5b5", 16, {10, 5}, {5, 5}, {0, 5}, {15, 1}},
+   {"r8g8b8", 24, {16, 8}, {8, 8}, {0, 8}, {0, 0}},
+   {"x8r8g8b8", 32, {16, 8}, {8, 8}, {0, 8}, {0, 0}},
+   {"a8r8g8b8", 32, {16, 8}, {8, 8}, {0, 8}, {24, 8}},
+   {"x8b8g8r8", 32, {0, 8}, {8, 8}, {16, 8}, {0, 0}},
+   {"a8b8g8r8", 32, {0, 8}, {8, 8}, {16, 8}, {24, 8}},
+   {"x2r10g10b10", 32, {20, 10}, {10, 10}, {0, 10}, {0, 0}},
+   {"a2r10g10b10", 32, {20, 10}, {10, 10}, {0, 10}, {30, 2}},
+};
+
+void alloc_mbi(void)
+{
+   void *mbi_memtop;
+
+   /*
+    * Place multiboot information at the top of
+    * 8 MB identity mapping space.
+    */
+   mbi_memtop = (void *)(LIN_VA_TO_PA(BASE_VA) + EARLY_MAP_SIZE);
+
+   mbi = (multiboot_info_t *)((ulong)mbi_memtop - sizeof(*mbi));
+   bzero(mbi, sizeof(*mbi));
+
+   mod = (multiboot_module_t *)((ulong)mbi - sizeof(*mod));
+   bzero(mod, sizeof(*mod));
+
+   cmdline_buf = (char *)mod - CMDLINE_BUF_SZ;
+   bzero(cmdline_buf, CMDLINE_BUF_SZ);
+}
+
+static void
+add_multiboot_mmap(u64 addr, u64 size, u32 type)
+{
+   /* Allocate a new multiboot_memory_map */
+   mmmap = (void *)(cmdline_buf - \
```

Comment:
This is hard to read: can't you just use the value of the previous `mmap` pointer instead of re-calculating the new position like this for every call? You can handle the first call by checking if `mmap` is NULL (I assume .bss is zeroed at this point). 

Also, `mmap` already has the right type. So, after the call where mmap is NULL and you initialize to point at `cmdline_buf - sizeof(multiboot_memory_map_t)`, for the following calls you can just do `mmap--` here. 

Makes sense?

Reply:
fixed!